agent/http.go

```diff
@@ -161,28 +160,55 @@ func (fs *redirectFS) Open(name string) (http.File, error) {
 	return file, err
 }
 
-type templatedIndexFS struct {
-	fs           http.FileSystem
-	templateVars func() map[string]interface{}
+type settingsInjectedIndexFS struct {
+	fs         http.FileSystem
+	UISettings map[string]interface{}
 }
 
-func (fs *templatedIndexFS) Open(name string) (http.File, error) {
+func (fs *settingsInjectedIndexFS) Open(name string) (http.File, error) {
 	file, err := fs.fs.Open(name)
 	if err != nil || name != "/index.html" {
 		return file, err
 	}
 
-	content, _ := ioutil.ReadAll(file)
+	content, err := ioutil.ReadAll(file)
+	if err != nil {
+		return nil, fmt.Errorf("failed reading index.html: %s", err)
+	}
 	file.Seek(0, 0)
-	t, err := template.New("fmtedindex").Parse(string(content))
+
+	// Replace the placeholder in the meta ENV with the actual UI config settings.
+	// Ember passes the ENV with URL encoded JSON in a meta tag. We are replacing
+	// a key and value that is the encoded version of `"CONSUL_UI_SETTINGS":"__CONSUL_UI_SETTINGS_GO_HERE__"`
+	// with a URL-encoded JSON blob representing the actual config.
+
+	// First built an escaped, JSON blob from the settings passed.
+	bs, err := json.Marshal(fs.UISettings)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed marshalling UI settings JSON: %s", err)
 	}
-	var out bytes.Buffer
-	if err := t.Execute(&out, fs.templateVars()); err != nil {
-		return nil, err
+	// We want to remove the first and last chars which will be the { and } since
+	// we are injecting these variabled into the middle of an existing object.
+	bs = bytes.Trim(bs, "{}")
```

Comment:
minor: I guess we could probably remove the need for this trim by making the replacement string include the `{}` and moving the comment around. Possibly even putting it into a separate file, although I'm not sure how easy that is.

Reply:
Hmm I don't think so - we are trimming the JSON value that is being inserted here.

The KV pair being matched appears somewhere in the middle of a large JSON blob that contains many other variables that ember sets already.

But I'm not sure what you meant about a separate file so perhaps I'm misunderstanding something?

Reply:
Oh ya, I didn't explain this well and thinking through it again my idea changed slightly.

Roughly I was thinking that some other file could set a variable to `nil`, and `environment.js` could use that variable as:

```
ENV = Object.assign({}, ENV, inject.CONSUL_INJECTED_UI_SETTINGS),
```

where `inject` is the name of the module that contains the other file (although I don't fully understand how modules work here).

Then the replacement could replace `inject.CONSUL_INJECTED_UI_SETTINGS`. The replacement string would not contain any escaped characters (I think), and the source of the variable wouldn't be replaced because it wouldn't include `inject.`.

That would remove the need to trim `{}` because the content would be an entire JSON object, not just the values, and it would make it much easier to associate these two strings because you could find or grep for them without having to escape things.

Reply:
Yeah there is certainly room to make this cleaner but I don't think what you're suggesting works. The `environment.js` is evaluated at Ember _build time_. The end result is the index.html that we store in assetFS which has this compiled, encoded JSON object in. We can't inject directly into the runtime JS without doing something more complicated like trying to intercept other JS file requests on the HTTP server and inject at a higher level where the JS has already been minified etc. This seems less bad for now and an improvement over what we were doing before at least!

As this is blocking getting something to work I'd prefer to defer a total refactor of how we pass variables to JS - there is already a big TODO in the JS side to change how we pass all of this since it should ideally be in the APP config section and handled in the `Application.initialize` rather than using `env()` anyway but that's too much to take on immediately.

Reply:
If we wanted to inject these things at runtime I think we would probably query them via the API, instead of injecting them into a file, right? I really like that idea, but I assumed there was some reason that wouldn't work. Do we need these settings in order to make API requests?

I think what I was suggesting isn't much different from what you have, but I don't know ember well enough to know if it is possible.